src/main/scala/vectorpipe/osm/internal/functions/package.scala

```diff
@@ -0,0 +1,636 @@
+              case None => connectSegments(t, lines :+ h)
+            }
+        }
+    }
+  }
+
+  private def connectSegments(segments: GenTraversable[LineString])(implicit geometryFactory: GeometryFactory): GenTraversable[LineString] =
+    connectSegments(segments.map(_.getCoordinateSequence).map(s => new VirtualCoordinateSequence(Seq(s))))
+      .map(geometryFactory.createLineString)
+
+  private def getHoles(poly: Polygon): List[CoordinateSequence] = {
+    for (i <- 0 until poly.getNumInteriorRing) yield
+      poly.getInteriorRingN(i).getCoordinateSequence
+  }.toList
+
+  private def dissolveRings(rings: GenTraversable[Polygon]): (Seq[Polygon], Seq[Polygon]) = {
+    if (!rings.isEmpty)
+      (rings.map(_.getExteriorRing).map(x => GeomFactory.factory.createPolygon(x.getCoordinateSequence)).toVector, rings.flatMap(getHoles).map(GeomFactory.factory.createPolygon).toVector)
+    else
+      (Vector.empty[Polygon], Vector.empty[Polygon])
+  }
+
+  // TODO this (and accompanying functions) doesn't belong here
+  def buildMultiPolygon(id: Long, version: Int, timestamp: Timestamp, types: Seq[Byte], roles: Seq[String], wkbs: Seq[Geometry]): Option[Geometry] = {
+    if (types.zip(wkbs).exists { case (t, g) => t == WayType && Option(g).isEmpty }) {
+      // bail early if null values are present where they should exist (members w/ type=way)
+      logger.debug(s"Incomplete relation: $id @ $version ($timestamp)")
+      None
+    } else {
+
+      val geoms = wkbs.map { geom =>
+        geom match {
+          case p: Polygon => Some(p.getExteriorRing())
+          case l: LineString => Some(l)
+          case null => None
+        }
+      }
+
+      val vertexCount = geoms.filter(_.isDefined).map(_.get).map(_.getNumPoints).sum
+
+      logger.warn(s"${vertexCount.formatted("%,d")} vertices from ${types.size} members in $id @ $version ($timestamp)")
+
+      val members: Seq[(String, LineString)] = roles.zip(geoms)
+        .filter(_._2.isDefined)
+        .map(x => (x._1, x._2.get))
+
+      val (complete, partial) = members.foldLeft((Vector.empty[Polygon], Vector.empty[LineString])) {
+        case ((c, p), (role, line: LineString)) =>
+          role match {
+            case "outer" if line.isClosed && line.getNumPoints >= 4 =>
+              (c :+ GeomFactory.factory.createPolygon(line.getCoordinateSequence), p)
+            case "outer" => (c, p :+ line)
+            case "inner" if line.isClosed && line.getNumPoints >= 4 =>
+              (c :+ GeomFactory.factory.createPolygon(line.getCoordinateSequence), p)
+            case "inner" => (c, p :+ line)
+            case "" if line.isClosed && line.getNumPoints >= 4 =>
+              (c :+ GeomFactory.factory.createPolygon(line.getCoordinateSequence), p)
+            case "" => (c, p :+ line)
+            case _ => (c, p)
+          }
+      }
+
+      try {
+        val rings = complete ++ formRings(partial.sortWith(_.getNumPoints > _.getNumPoints))
+        val preparedRings = rings
```

Comment:
One of the changes I made from @mojodna's original code was changing `preparedRings` from a `PreparedGeometry` to just a `Geometry`. I'm not familiar with `PreparedGeometries`, so I'm not sure if the change has a significant impact or not. This is the original code: https://github.com/azavea/osmesa/blob/master/src/ingest/src/main/scala/osmesa/functions/osm/package.scala#L509

Reply:
Using prepared geometries is actually an important optimization. A prepared geometry is one in which predicates like 'intersects' runs much faster than the unprepared version.

As a general rule, I'd recommend against taking out or changing functionality that isn't well understood, without consulting with the author or getting a clear understanding of why the choices were made - those choices could come from hours of debugging and optimizing, or they could be mistakes, but it's important to know which. Not knowing is a common way I've seen regressions happen!

Thanks for bringing that point up.

Reply:
I'll take a look at these. I removed the preparedness while switching to use JTS for dissolving rings.

Reply:
I see. Next time I will look into the type more and/or ask someone for their opinion. I have since reverted my changes and now `PreparedGeometries` are used. Thanks for explaining what those types are.

Reply:
Ah, I didn't realize that I left the `PreparedGeometries` in. Great.

Reply:
@mojodna are prepared geoms not required here then?

Reply:
They are required. I thought I'd removed them on the OSMesa side when switching to JTS after not reading @jbouffard's comment about having changed them here.

Reply:
Ah, gotcha.